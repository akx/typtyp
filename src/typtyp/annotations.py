import dataclasses


@dataclasses.dataclass(frozen=True)
class Comment:
    comment: str
