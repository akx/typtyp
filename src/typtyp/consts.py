import collections

COLLECTION_ORIGINS = (
    list,
    collections.deque,
    set,
    frozenset,
    collections.abc.Iterable,
    collections.abc.Iterator,
    collections.abc.Sequence,
)

MAPPING_ORIGINS = (
    dict,
    collections.defaultdict,
    collections.abc.Mapping,
    collections.abc.MutableMapping,
    collections.OrderedDict,
)
