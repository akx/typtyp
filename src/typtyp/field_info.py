import dataclasses
from typing import TypedDict


@dataclasses.dataclass(frozen=True)
class FieldInfo:
    name: str
    type: type
    doc: str | None = None


class FieldInfoDict(TypedDict, total=False):
    name: str
    type: type
    doc: str | None
