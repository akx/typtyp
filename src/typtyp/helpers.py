from typing import Hashable, Iterable, TypeVar

THashable = TypeVar("THashable", bound=Hashable)


def unique_in_order(values: Iterable[THashable]) -> Iterable[THashable]:
    seen = set()
    for v in values:
        if v in seen:
            continue
        seen.add(v)
        yield v
