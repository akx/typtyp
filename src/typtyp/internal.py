import dataclasses


@dataclasses.dataclass(frozen=True)
class FieldInfo:
    name: str
    type: type
    doc: str | None = None
