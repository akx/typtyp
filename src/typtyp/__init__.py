from typtyp.field_info import FieldInfo, FieldInfoDict
from typtyp.type_configuration import TypeConfiguration
from typtyp.world import World

__all__ = [
    "FieldInfo",
    "FieldInfoDict",
    "TypeConfiguration",
    "World",
]
