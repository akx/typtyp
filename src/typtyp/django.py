from __future__ import annotations

import datetime
import uuid
from decimal import Decimal
from typing import Annotated, Any, Iterable, Literal, Optional, Union

from django.db import models

from typtyp.annotations import Comment
from typtyp.internal import FieldInfo

STRINGLIKE_FIELDS = (
    models.CharField,
    models.EmailField,
    models.GenericIPAddressField,
    models.SlugField,
    models.TextField,
    models.URLField,
)

SIMPLE_FIELD_TYPES = {
    models.BinaryField: bytes,
    models.BooleanField: bool,
    models.DateField: datetime.date,
    models.DateTimeField: datetime.datetime,
    models.DecimalField: Decimal,
    models.DurationField: str,
    models.FileField: Annotated[str, Comment("file field")],  # TODO: Handle more specific types?
    models.FloatField: float,
    models.ImageField: Annotated[str, Comment("image field")],  # TODO: Handle more specific types?
    models.IntegerField: int,
    models.JSONField: Any,
    models.TimeField: datetime.time,
    models.UUIDField: uuid.UUID,
}


def get_referred_model_type(field: models.Field, model_type: type[models.Model]):
    if field.related_model == "self":
        return model_type  # self-referential relation
    return field.related_model


def get_django_field_type(field: models.Field, *, model_type: type[models.Model]) -> tuple[type, str | None] | None:
    if flatchoices := getattr(field, "flatchoices", None):
        return (Union[*(Literal[c[0]] for c in flatchoices)], None)
    if isinstance(field, STRINGLIKE_FIELDS):
        return (str, None)
    for field_type, typ in SIMPLE_FIELD_TYPES.items():
        if isinstance(field, field_type):
            return (typ, None)
    if isinstance(field, (models.ManyToManyRel, models.ManyToOneRel)):
        # Skip reverse relations for now.
        return None
    if isinstance(field, models.ManyToManyField):
        return (list[get_referred_model_type(field, model_type)], "many-to-many relation")
    if isinstance(field, models.OneToOneField):
        return (get_referred_model_type(field, model_type), "one-to-one relation")
    if isinstance(field, models.ForeignKey):
        return (get_referred_model_type(field, model_type), "foreign key relation")
    raise NotImplementedError(f"Unsupported Django field type ({type(field).__name__}): {field!r}")


def is_django_model(t) -> bool:
    """
    Infallibly checks if a type is a Django model. Never throws.
    """
    try:
        return issubclass(t, models.Model)
    except Exception:
        return False


def get_model_fields(model_type: type[models.Model]) -> Iterable[FieldInfo]:
    for field in sorted(model_type._meta.get_fields(include_hidden=True), key=lambda field: field.name):
        typ = get_django_field_type(field, model_type=model_type)
        if typ is None:
            continue
        typ, comment = typ
        if field.null:
            typ = Optional[typ]
        if comment:
            typ = Annotated[typ, Comment(comment=comment)]
        yield FieldInfo(field.name, typ, doc=str(field.help_text) if field.help_text else None)
