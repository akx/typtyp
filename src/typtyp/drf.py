from __future__ import annotations

import datetime
import inspect
import uuid
from decimal import Decimal
from typing import Annotated, Any, Iterable, Literal, Union

from rest_framework import fields, relations, serializers

from typtyp.annotations import Comment
from typtyp.field_info import FieldInfo

STRINGLIKE_FIELDS = (
    fields.CharField,
    fields.EmailField,
    fields.RegexField,
    fields.SlugField,
    fields.URLField,
)

SIMPLE_FIELD_TYPES = {
    fields.BooleanField: bool,
    fields.DateField: datetime.date,
    fields.DateTimeField: datetime.datetime,
    fields.DecimalField: Decimal,
    fields.DurationField: str,
    fields.FileField: Annotated[str, Comment("file field")],  # TODO: Handle more specific types?
    fields.FloatField: float,
    fields.ImageField: Annotated[str, Comment("image field")],  # TODO: Handle more specific types?
    fields.IntegerField: int,
    fields.JSONField: Any,
    fields.TimeField: datetime.time,
    fields.ReadOnlyField: Annotated[  # TODO: can we dig into these?
        Any,
        Comment(comment="read only field, type unknown"),
    ],
    fields.UUIDField: uuid.UUID,
    relations.HyperlinkedRelatedField: str,
    relations.PrimaryKeyRelatedField: Annotated[  # TODO: Handle more specific types?
        Any,
        Comment(comment="primary key related field"),
    ],
}


def is_drf_serializer(t) -> bool:
    """
    Infallibly checks if a type is a DRF serializer. Never throws.
    """
    try:
        return issubclass(t, serializers.Serializer)
    except Exception:  # pragma: no cover
        return False


def get_drf_field_type(field: fields.Field) -> type:
    if isinstance(field, serializers.DictField):
        return dict[Any, get_drf_field_type(field.child)]
    if isinstance(field, fields._UnvalidatedField):
        return Any
    if isinstance(field, fields.ListField):
        return list[get_drf_field_type(field.child)]
    if isinstance(field, serializers.Serializer):
        return field.__class__
    if isinstance(field, serializers.ListSerializer):
        return list[get_drf_field_type(field.child)]
    if isinstance(field, fields.ChoiceField):
        return Union[*(Literal[choice] for choice in field.choices)]
    if isinstance(field, serializers.SerializerMethodField):
        method = getattr(field.parent, field.method_name)
        rtype = inspect.get_annotations(method).get("return", Any)
        if isinstance(rtype, str):
            # If the return type is a string, assume it's a forward reference
            return Annotated[
                Any,
                Comment(comment=f"forward reference: {rtype}"),
            ]
        return rtype
    if isinstance(field, STRINGLIKE_FIELDS):
        return str
    for field_type, typ in SIMPLE_FIELD_TYPES.items():
        if isinstance(field, field_type):
            return typ
    raise NotImplementedError(f"Unsupported DRF field type ({type(field).__name__}): {field!r}")  # pragma: no cover


def get_serializer_fields(ser_type: type[serializers.Serializer]) -> Iterable[FieldInfo]:
    context = {"request": None}
    for name, field in sorted(ser_type(context=context).fields.items(), key=lambda item: item[0]):
        yield FieldInfo(
            name=name,
            type=get_drf_field_type(field),
            doc=str(field.help_text) if field.help_text else None,
        )
