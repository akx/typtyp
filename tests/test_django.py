import pytest

pytest.importorskip("django")

from django.contrib.auth.models import Group, Permission, User
from django.contrib.contenttypes.models import ContentType

import typtyp
from tests.django_interop.models import PetEmergencyContact, PetMedication, PetSittingGig


def test_django(snapshot):
    w = typtyp.World()

    w.add_many(
        (
            ContentType,
            Group,
            Permission,
            PetEmergencyContact,
            PetMedication,
            PetSittingGig,
            User,
        ),
    )
    assert w.get_typescript() == snapshot
