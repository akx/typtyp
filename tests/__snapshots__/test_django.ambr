# serializer version: 1
# name: test_django
  '''
  /** ContentType(id, app_label, model) */
  export interface ContentType {
  app_label: string
  id: number
  model: string
  }
  /**
   * Groups are a generic way of categorizing users to apply permissions, or
   * some other label, to those users. A user can belong to any number of
   * groups.
   * 
   * A user in a group automatically has all the permissions granted to that
   * group. For example, if the group 'Site editors' has the permission
   * can_edit_home_page, any user in that group will have that permission.
   * 
   * Beyond permissions, groups are a convenient way to categorize users to
   * apply some label, or extended functionality, to them. For example, you
   * could create a group 'Special users', and you could write code that would
   * do special things to those users -- such as giving them access to a
   * members-only portion of your site, or sending them members-only email
   * messages.
   */
  export interface Group {
  id: number
  name: string
  permissions: (Permission)[] /* many-to-many relation */
  }
  /**
   * The permissions system provides a way to assign permissions to specific
   * users and groups of users.
   * 
   * The permission system is used by the Django admin site, but may also be
   * useful in your own code. The Django admin site uses permissions as follows:
   * 
   *     - The "add" permission limits the user's ability to view the "add" form
   *       and add an object.
   *     - The "change" permission limits a user's ability to view the change
   *       list, view the "change" form and change an object.
   *     - The "delete" permission limits the ability to delete an object.
   *     - The "view" permission limits the ability to view an object.
   * 
   * Permissions are set globally per type of object, not per specific object
   * instance. It is possible to say "Mary may change news stories," but it's
   * not currently possible to say "Mary may change news stories, but only the
   * ones she created herself" or "Mary may only change news stories that have a
   * certain status or publication date."
   * 
   * The permissions listed above are automatically created for each model.
   */
  export interface Permission {
  codename: string
  content_type: ContentType /* foreign key relation */
  id: number
  name: string
  }
  /** Because pets always have drama at 3 AM */
  export interface PetEmergencyContact {
  /** When they can actually help */
  availability: string
  gig: PetSittingGig /* foreign key relation */
  id: number
  name: string
  phone: string
  /** Aunt, neighbor, fellow pet parent, etc. */
  relationship: string
  }
  /** For when pets have more prescriptions than their owners */
  export interface PetMedication {
  /** Hidden in food, wrestling match, or magic trick */
  administration_method: string
  dosage: string
  /** How often to drug the pet */
  frequency: string
  gig: PetSittingGig /* foreign key relation */
  id: number
  medication_name: string
  }
  /**
   * A comprehensive model for a pet sitting business that somehow needs
   * to track an absurd amount of information about each gig.
   */
  export interface PetSittingGig {
  /** Emergency reinforcement */
  backup_sitter: User | null /* foreign key relation */
  created_at: ISO8601Date
  /** What this chaos costs */
  daily_rate: number
  difficulty_level: "easy" | "moderate" | "hard" | "expert" | "impossible"
  /** When things go sideways */
  emergency_contact_info: string
  /** When freedom returns */
  end_date: ISO8601Date
  /** Auto-incrementing gig number */
  gig_number: number
  /** Please don't set this off */
  house_alarm_code: number | null
  /** Which key opens what */
  house_key_photo: string /* file field */
  /** Encrypted WiFi password */
  house_wifi_password: unknown /* bytes */
  id: UUID
  /** Houdini tendencies? */
  is_escape_artist: boolean
  /** Are we dealing with surprises? */
  is_house_trained: boolean
  /** When they last conned someone into food */
  last_fed: ISO8601Date | null
  /** Am I now a pet pharmacist? */
  needs_medication: boolean
  /** Size of the furry army */
  number_of_pets: number
  /** Scale of 1-10, where 10 is 'calls every hour' */
  owner_anxiety_level: number
  /** For sending cute pet pics */
  owner_email: string
  /** Am I being watched? */
  owner_has_security_cameras: boolean
  /** Novel-length instructions from neurotic pet parents */
  owner_notes: string
  payment_status: "pending" | "partial" | "paid" | "overdue"
  /** Because of course they have one */
  pet_instagram: string
  /** Do I need to maintain their brand? */
  pet_is_social_media_famous: boolean
  /** What do we call this furry overlord? */
  pet_name: string
  /** Their street name */
  pet_nickname: string
  /** Mugshot for identification */
  pet_photo: string /* file field */
  pet_species: "dog" | "cat" | "bird" | "fish" | "hamster" | "snake" | "turtle" | "other"
  /** In pounds (for dosing purposes) */
  pet_weight: number
  /** Where the good stuff is stashed */
  secret_treat_hiding_spots: string
  /** For checking in remotely */
  security_camera_ip: string | null
  /** The brave soul taking this on */
  sitter: User /* foreign key relation */
  /** URL-friendly identifier */
  slug: string
  /** Structured chaos: feeding times, quirks, emergency protocols */
  special_instructions: unknown /* any */
  /** When the adventure begins */
  start_date: ISO8601Date
  /** Exact moment of responsibility transfer */
  start_time: ISO8601Time
  /** How long they drag you around the block */
  typical_walk_duration: string
  updated_at: ISO8601Date
  /** Proof of shots */
  vaccination_records: string /* file field */
  /** Dr. Whiskers' online presence */
  vet_website: string
  }
  /**
   * Users within the Django authentication system are represented by this
   * model.
   * 
   * Username and password are required. Other fields are optional.
   */
  export interface User {
  date_joined: ISO8601Date
  email: string
  first_name: string
  /** The groups this user belongs to. A user will get all permissions granted to each of their groups. */
  groups: (Group)[] /* many-to-many relation */
  id: number
  /** Designates whether this user should be treated as active. Unselect this instead of deleting accounts. */
  is_active: boolean
  /** Designates whether the user can log into this admin site. */
  is_staff: boolean
  /** Designates that this user has all permissions without explicitly assigning them. */
  is_superuser: boolean
  last_login: ISO8601Date | null
  last_name: string
  password: string
  /** Specific permissions for this user. */
  user_permissions: (Permission)[] /* many-to-many relation */
  /** Required. 150 characters or fewer. Letters, digits and @/./+/-/_ only. */
  username: string
  }
  export type ISO8601Date = string
  export type ISO8601Time = string
  export type UUID = string
  
  '''
# ---
